File: bin/forged.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/wait.h>
#include <sys/stat.h>

#define LOG(msg) printf("\033[1;35m[FORGED]\033[0m %s\n", msg)
#define RUN(argv) do { \
    pid_t p = fork(); \
    if (p == 0) { execvp(argv[0], (char *const *)argv); exit(1); } \
    else { int s; waitpid(p, &s, 0); if (WEXITSTATUS(s)) exit(1); } \
} while (0)

int dir_exists(const char *path) {
    struct stat st;
    return stat(path, &st) == 0 && S_ISDIR(st.st_mode);
}

int file_exists(const char *path) {
    struct stat st;
    return stat(path, &st) == 0 && !S_ISDIR(st.st_mode);
}

int main(int argc, char **argv) {
    if (argc < 3) {
        LOG("Usage: forged [-B pkg] [-G repo] [-GR name] [-R pkg]");
        return 1;
    }

    char *op = argv[1];

    if (!strcmp(op, "-B")) {
        char *pkg = argv[2];
        LOG("Searching smithies...");
        FILE *fp = popen("pacman -Ss ^${pkg}$", "r");
        char official[1024] = {0};
        if (fp) fgets(official, sizeof(official), fp);
        pclose(fp);
        char *official_pkg = strtok(official, " ");

        fp = popen("yay -Ssa ${pkg}", "r");
        char aur[1024] = {0};
        if (fp) fgets(aur, sizeof(aur), fp);
        pclose(fp);
        char *aur_pkg = strtok(strstr(aur, "/") ? strstr(aur, "/") + 1 : aur, " ");

        if (!official_pkg && !aur_pkg) {
            LOG("No smithy found");
            return 1;
        }

        printf("[FORGED] Found in %s smithy(ies):\n", official_pkg && aur_pkg ? "two" : "one");
        int count = 1;
        if (official_pkg) printf("  [%d] %s - official repos\n", count++, official_pkg);
        if (aur_pkg) printf("  [%d] %s - AUR\n", count, aur_pkg);

        char choice[16];
        printf("[FORGED] Pick a source (ENTER = official): ");
        fgets(choice, sizeof(choice), stdin);
        char *chosen = (choice[0] == '\n' && official_pkg) ? official_pkg : (choice[0] == '2' ? aur_pkg : official_pkg);

        LOG("Installing...");
        const char *cmd[] = {"yay", "-S", "--noconfirm", "--needed", chosen, NULL};
        RUN(cmd);
        LOG("Installed");
    } else if (!strcmp(op, "-G")) {
        for (int i = 2; i < argc; i++) {
            char *target = argv[i];
            char name[256];
            strcpy(name, strrchr(target, '/') ? strrchr(target, '/') + 1 : target);
            char *dotgit = strstr(name, ".git");
            if (dotgit) *dotgit = 0;

            LOG("Forging from source...");
            system("mkdir -p ~/forged/git");
            chdir("~/forged/git");
            char rmcmd[512];
            snprintf(rmcmd, sizeof(rmcmd), "rm -rf %s", name);
            system(rmcmd);
            const char *clone[] = {"git", "clone", "--depth=1", target, name, NULL};
            RUN(clone);
            chdir(name);

            if (file_exists("meson.build")) {
                LOG("Meson detected");
                const char *meson[] = {"meson", "setup", "build", NULL};
                RUN(meson);
                chdir("build");
                const char *ninja[] = {"ninja", NULL};
                RUN(ninja);
                const char *install[] = {"sudo", "ninja", "install", NULL};
                RUN(install);
            } else if (file_exists("CMakeLists.txt")) {
                LOG("CMake detected");
                const char *cmake[] = {"cmake", "-Bbuild", "-DCMAKE_INSTALL_PREFIX=/usr", NULL};
                RUN(cmake);
                const char *build[] = {"cmake", "--build", "build", "-j$(nproc)", NULL};
                RUN(build);
                const char *install[] = {"sudo", "cmake", "--install", "build", NULL};
                RUN(install);
            } else if (file_exists("Makefile") || file_exists("makefile")) {
                LOG("Make detected");
                const char *make[] = {"make", "-j$(nproc)", NULL};
                RUN(make);
                const char *install[] = {"sudo", "make", "install", NULL};
                RUN(install);
            } else if (file_exists("PKGBUILD")) {
                LOG("PKGBUILD detected");
                const char *makepkg[] = {"makepkg", "-si", "--noconfirm", NULL};
                RUN(makepkg);
            } else {
                LOG("No build system - shell");
                const char *bash[] = {"bash", NULL};
                RUN(bash);
            }
            LOG("Forged from source");
        }
    } else if (!strcmp(op, "-GR")) {
        char *name = argv[2];
        char path[512];
        snprintf(path, sizeof(path), "~/forged/git/%s", name);
        if (!dir_exists(path)) {
            LOG("Nothing forged");
            return 1;
        }
        LOG("Smelting...");
        char rmcmd[512];
        snprintf(rmcmd, sizeof(rmcmd), "rm -rf %s", path);
        system(rmcmd);
        LOG("Removed");
    } else if (!strcmp(op, "-R")) {
        LOG("Removing...");
        char **cmd = malloc((argc + 5) * sizeof(char*));
        cmd[0] = "sudo"; cmd[1] = "pacman"; cmd[2] = "-Rns"; cmd[3] = "--noconfirm";
        for (int i = 2; i < argc; i++) cmd[i + 2] = argv[i];
        cmd[argc + 2] = NULL;
        RUN((const char **)cmd);
        free(cmd);
        LOG("Removed");
    } else {
        // Forward other flags to yay
        LOG("Forwarding to yay...");
        char **cmd = malloc((argc + 1) * sizeof(char*));
        cmd[0] = "yay";
        for (int i = 1; i < argc; i++) cmd[i] = argv[i];
        cmd[argc] = NULL;
        RUN((const char **)cmd);
        free(cmd);
    }

    LOG("Steel is tempered");
    return 0;
}

File: bin/forged.go
// forged.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

const version = "2.1.0"

const (
	reset   = "\033[0m"
	bold    = "\033[1m"
	green   = "\033[32m"
	yellow  = "\033[33m"
	red     = "\033[31m"
	cyan    = "\033[36m"
	blue    = "\033[34m"
	magenta = "\033[35m"
)

func main() {
	if len(os.Args) < 2 {
		usage()
	}

	switch os.Args[1] {
	case "-B", "--bin":
		handleBin(os.Args[2:])
	case "-G", "--git":
		handleGit(os.Args[2:])
	case "-Syu", "--sync-upgrade":
		fullUpgrade()
	case "-v", "--version":
		fmt.Printf("%s forged %s — the Archforged package manager\n", green+"forged"+reset, cyan+version+reset)
	default:
		usage()
	}
}

func usage() {
	fmt.Printf(`
%s%s%s
%s

Usage:
  %s -B <package>           → install from repos or AUR
  %s -G <url>               → clone & auto-build git project
  %s -Syu                   → full system upgrade
  %s -v                     → show version

Examples:
  %s hyprland
  %s visual-studio-code-bin
  %s https://github.com/hyprwm/hyprpaper
`, magenta+bold+"forged"+reset, reset, cyan+version+reset, yellow+"The future of ricing"+reset,
		green+"forged"+reset, green+"forged"+reset, green+"forged"+reset, green+"forged"+reset,
		green+"forged -B"+reset, green+"forged -B"+reset, green+"forged -G"+reset)
	os.Exit(0)
}

func handleBin(args []string) {
	if len(args) == 0 {
		fmt.Printf("%s package name required\n", red+"error"+reset)
		os.Exit(1)
	}
	pkg := args[0]
	fmt.Printf("%s Installing %s...\n", blue+"install"+reset, yellow+pkg+reset)

	// Try official repos
	if run("pacman", "-Si", pkg) == nil {
		if err := run("sudo", "pacman", "-S", "--noconfirm", pkg); err != nil {
			fmt.Printf("%s failed to install %s from official repos\n", red+"error"+reset, yellow+pkg+reset)
			os.Exit(1)
		}
		fmt.Printf("%s %s installed from official repositories\n", green+"success"+reset, pkg)
		return
	}

	// AUR fallback
	var helperCmd string
	if hasYay() {
		helperCmd = "yay"
	} else if hasParu() {
		helperCmd = "paru"
	} else {
		fmt.Printf("%s Installing yay first...\n", yellow+"warning"+reset)
		installYay()
		helperCmd = "yay"
	}

	if err := run(helperCmd, "-S", "--noconfirm", pkg); err != nil {
		fmt.Printf("%s failed to install %s from AUR\n", red+"error"+reset, yellow+pkg+reset)
		os.Exit(1)
	}

	fmt.Printf("%s %s installed from AUR\n", green+"success"+reset, pkg)
}

func handleGit(args []string) {
	if len(args) == 0 {
		fmt.Printf("%s git URL required\n", red+"error"+reset)
		os.Exit(1)
	}
	url := args[0]
	name := strings.TrimSuffix(filepath.Base(url), ".git")
	dir := fmt.Sprintf("/tmp/forged-%s-%d", name, time.Now().Unix())

	fmt.Printf("%s Cloning %s → %s\n", cyan+"download"+reset, yellow+name+reset, dir)
	if err := run("git", "clone", "--depth=1", url, dir); err != nil {
		fmt.Printf("%s git clone failed for %s\n", red+"error"+reset, yellow+url+reset)
		os.Exit(1)
	}
	os.Chdir(dir)

	switch {
	case fileExists("PKGBUILD"):
		buildPkgbuild(name)
	case fileExists("meson.build"):
		buildMeson(name)
	case fileExists("CMakeLists.txt"):
		buildCMake(name)
	case fileExists("package.json"):
		buildNode(name)
	case fileExists("pom.xml"):
		buildMaven(name)
	case fileExists("build.gradle"), fileExists("gradlew"):
		buildGradle(name)
	case fileExists("configure"), fileExists("Makefile.am"):
		buildAutotools(name)
	case fileExists("go.mod"):
		buildGo(name)
	default:
		fmt.Printf("%s No build system detected — dropping into shell\n", yellow+"warning"+reset)
		run("bash")
	}
}

func exitOnErr(err error, name string, step string) {
	if err != nil {
		fmt.Printf("%s %s failed for %s\n", red+"error"+reset, step, yellow+name+reset)
		os.Exit(1)
	}
}

func buildPkgbuild(name string) {
	fmt.Printf("%s PKGBUILD → makepkg -si\n", cyan+"build"+reset)
	exitOnErr(run("makepkg", "-si", "--noconfirm"), name, "makepkg")
}
func buildMeson(name string) {
	fmt.Printf("%s Meson → ninja install\n", cyan+"build"+reset)
	exitOnErr(run("meson", "setup", "build", "--prefix=/usr"), name, "meson setup")
	exitOnErr(run("ninja", "-C", "build"), name, "ninja build")
	exitOnErr(run("sudo", "ninja", "-C", "build", "install"), name, "ninja install")
}
func buildCMake(name string) {
	fmt.Printf("%s CMake → make install\n", cyan+"build"+reset)
	exitOnErr(run("cmake", "-Bbuild", "-DCMAKE_INSTALL_PREFIX=/usr"), name, "cmake")
	exitOnErr(run("make", "-Cbuild", "-j$(nproc)"), name, "make")
	exitOnErr(run("sudo", "make", "-Cbuild", "install"), name, "make install")
}
func buildNode(name string) {
	fmt.Printf("%s Node.js → npm install && build\n", cyan+"build"+reset)
	exitOnErr(run("npm", "install"), name, "npm install")
	if fileExists("webpack.config.js") || fileExists("vite.config.js") {
		exitOnErr(run("npm", "run", "build"), name, "npm run build")
	}
}
func buildMaven(name string) {
	fmt.Printf("%s Maven → mvn install\n", cyan+"build"+reset)
	exitOnErr(run("mvn", "clean", "install"), name, "mvn install")
}
func buildGradle(name string) {
	fmt.Printf("%s Gradle → ./gradlew build\n", cyan+"build"+reset)
	if fileExists("gradlew") {
		exitOnErr(run("chmod", "+x", "gradlew"), name, "chmod gradlew")
		exitOnErr(run("./gradlew", "build"), name, "./gradlew build")
	} else {
		exitOnErr(run("gradle", "build"), name, "gradle build")
	}
}
func buildAutotools(name string) {
	fmt.Printf("%s Autotools → ./configure && make install\n", cyan+"build"+reset)
	if !fileExists("configure") {
		exitOnErr(run("autoreconf", "-fi"), name, "autoreconf")
	}
	exitOnErr(run("./configure", "--prefix=/usr"), name, "./configure")
	exitOnErr(run("make", "-j$(nproc)"), name, "make")
	exitOnErr(run("sudo", "make", "install"), name, "make install")
}
func buildGo(name string) {
	fmt.Printf("%s Go → go build && install\n", cyan+"build"+reset)
	exitOnErr(run("go", "build", "."), name, "go build")
	// **BUG FIX**: Install './{name}' (the binary you just built), not os.Args[0]
	exitOnErr(run("sudo", "install", "./"+name, "/usr/local/bin/"), name, "go install")
}

func fullUpgrade() {
	fmt.Printf("%s Upgrading system + AUR...\n", blue+"upgrade"+reset)
	exitOnErr(run("sudo", "pacman", "-Syu", "--noconfirm"), "system", "pacman upgrade")
	if hasYay() {
		exitOnErr(run("yay", "-Syu", "--noconfirm"), "AUR", "yay upgrade")
	} else if hasParu() {
		exitOnErr(run("paru", "-Syu", "--noconfirm"), "AUR", "paru upgrade")
	}
	fmt.Printf("%s System fully upgraded!\n", green+"success"+reset)
}

func run(name string, args ...string) error {
	cmd := exec.Command(name, args...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	cmd.Stdin = os.Stdin
	return cmd.Run()
}

func fileExists(p string) bool { _, err := os.Stat(p); return err == nil }
func hasYay() bool            { return exec.Command("which", "yay").Run() == nil }
func hasParu() bool           { return exec.Command("which", "paru").Run() == nil }

func installYay() {
	if err := run("git", "clone", "https://aur.archlinux.org/yay.git", "/tmp/yay"); err != nil {
		fmt.Printf("%s failed to clone yay\n", red+"error"+reset)
		os.Exit(1)
	}
	os.Chdir("/tmp/yay")
	if err := run("makepkg", "-si", "--noconfirm"); err != nil {
		fmt.Printf("%s failed to build yay\n", red+"error"+reset)
		os.Exit(1)
	}
}
